src/dfcx_scrapi/core/test_cases.py

```diff
@@ -430,3 +432,149 @@ def calculate_coverage(self, coverage_type: int, agent_id: str = None):
         )
         response = client.calculate_coverage(request)
         return response
+
+    def _convert_flow(self, flow_id, flows_map):
+        """Gets a flow display name from a flow ID"""
+        if flow_id.split("/")[-1] == "-":
+            return ""
+        # flow_id_converted = str(agent_id) + '/flows/' + str(flow_id)
+        if flow_id in flows_map:
+            return flows_map[flow_id]
+        # TODO: Should throw error instead of returning default
+        return "Default Start Flow"
+
+    # Note that flow id includes agent, normally...
+    def _convert_page(self, page_id, flow_id, pages_map):
+        """Gets a page display name from a page and flow ID"""
+        if page_id == "END_SESSION":
+            return "End Session"
+        elif page_id == "END_FLOW":
+            return "End Flow"
+        elif page_id == "START_PAGE":
+            return "Start"
+        page_id_converted = str(flow_id) + "/pages/" + str(page_id)
+        if flow_id in pages_map:
+            if page_id_converted in pages_map[flow_id]:
+                return pages_map[flow_id][page_id_converted]
+            else:
+                # TODO: Should throw error instead of returning default
+                return "Start"
+        print("Flow not found")
+        # TODO: Should throw error, but returning this probably will anyway
+        return "Invalid"
+
+    def get_test_case_results_df(self, agent_id=None, retest_all=False):
+        """Gets the test case results for this agent,
+        and generates a dataframe with their details.
+        Any tests without a result will be run in a batch.
+
+        Args:
+          agent_id: required only if not set when initializing this class
+          retest_all: if true, all test cases are re-run,
+            regardless of whether or not they had a result
+
+        Returns:
+          DataFrame of test case results for this agent, with columns:
+            display_name, id, short_id (excluding agent ID),
+            tags (comma-separated string), creation_time,
+            start_flow, start_page, passed, test_time
+        """
+        if agent_id:
+            self.agent_id = agent_id
+
+        dfcx_flows = Flows(creds=self.creds, agent_id=self.agent_id)
+        dfcx_pages = Pages(creds=self.creds)
+        flows_map = dfcx_flows.get_flows_map(agent_id=self.agent_id)
+        pages_map = {}
+        for flow_id in flows_map.keys():
+            pages_map[flow_id] = dfcx_pages.get_pages_map(flow_id=flow_id)
+
+        test_case_results = self.list_test_cases(self.agent_id)
+        retest = []
+        retest_names = []
+
+        display_names = []
+        ids = []
+        short_ids = []
+        tags = []
+        creation_times = []
+        flows = []
+        pages = []
+        test_results = []
+        test_times = []
+        passed = []
+
+        for response in test_case_results:
+            # Collect untested cases to be retested
```

Comment:
@SeanScripts If you have a commented section in a method where you are describing what that section does, you can just make that entire section a new method.

This will clean up your main method and provide better debug tracking later.

Reply:
I moved this section to a new function and had it return a single-row dataframe for each test case response, which avoided initializing all the empty lists above. I kept the check for test cases that needed to be retested since it was just one condition.